pintc/src/predicate/optimize/dead_code_elimination.rs

```diff
@@ -147,3 +151,169 @@ pub(crate) fn dead_select_elimination(contract: &mut Contract) {
         }
     }
 }
+
+/// Replace all uses of, and remove all duplicate Variable declarations
+///
+/// Any variable declaration that contains an equivalent expression is removed, regardless of span
+///
+/// Ex.
+/// let a = x + 1;
+/// let b = x + 1;
+/// let d = b + 1;
+/// becomes
+/// let a = x + 1;
+/// let d = a + 1;
+pub(crate) fn duplicate_variable_elimination(contract: &mut Contract) {
+    for pred_key in contract.preds.keys().collect::<Vec<_>>() {
+        let pred_variables = contract
+            .preds
+            .get(pred_key)
+            .expect("pred guaranteed to exist")
+            .variables()
+            .collect::<Vec<_>>();
+
+        // gather all variable declarations with identical expressions
+        let mut dupe_var_decls: Vec<(
+            VariableKey, /* original */
+            VariableKey, /* duplicate */
+        )> = vec![];
+
+        for (i, (var_key, var)) in pred_variables.iter().enumerate() {
+            // avoid double checking any variable that has already been marked as a duplicate
+            if dupe_var_decls
+                .iter()
+                .any(|(original_var_key, dupe_var_key)| {
+                    original_var_key == var_key || dupe_var_key == var_key
+                })
+            {
+                continue;
+            };
+
+            for (subsequent_var_key, subsequent_var) in pred_variables.iter().skip(i + 1) {
+                if let Some(pred) = contract.preds.get(pred_key) {
```

Comment:
So you're checking if `contract.preds.get(pred_key)` exists multiple times in this file. If you want, you can do this check at the beginning only once (right after `for pred_key in contract.preds.keys().collect::<Vec<_>>() {`)